include/zephyr/drivers/stepper.h

```diff
@@ -0,0 +1,227 @@
+				      const int32_t value);
+
+/**
+ * @typedef stepper_get_attr_t
+ * @brief Get the value of a certain attribute in stepper controller
+ */
+typedef int32_t (*stepper_get_attr_t)(const struct device *dev, enum stepper_attribute,
+				      int32_t *value);
+
+__subsystem struct stepper_api {
+	stepper_enable_t stepper_enable;
+	stepper_set_steps_to_move_t stepper_set_steps_to_move;
+	stepper_set_velocity_t stepper_set_velocity;
+	stepper_set_micro_step_res_t stepper_set_micro_step_res;
+	stepper_set_attr_t stepper_set_attribute;
+	stepper_get_attr_t stepper_get_attribute;
+};
+
+/**
+ * @brief Enable or Disable Motor Controller
+ * @param dev pointer to the stepper motor controller instance
+ * @param enable Input enable or disable motor controller
+ * @return <0 Error during Enabling
+ *          0 Success
+ */
+__syscall int32_t stepper_enable(const struct device *dev, bool enable);
+
+static inline int32_t z_impl_stepper_enable(const struct device *dev, bool enable)
+{
+	const struct stepper_api *api = (const struct stepper_api *)dev->api;
+
+	return api->stepper_enable(dev, enable);
+}
+
+/**
+ * @brief Set the steps to be moved from the current position i.e. relative movement
+ * @param dev pointer to the stepper motor controller instance
+ * @param steps target steps to be moved from the current position
+ * @return <0 Error during Enabling
+ *          0 Success
+ */
+__syscall int32_t stepper_set_steps_to_move(const struct device *dev, int32_t steps);
+
+static inline int32_t z_impl_stepper_set_steps_to_move(const struct device *dev, int32_t steps)
+{
+	const struct stepper_api *api = (const struct stepper_api *)dev->api;
+
+	return api->stepper_set_steps_to_move(dev, steps);
+}
+
+/**
+ * @brief Set the target velocity to be reached by the motor
+ * @param dev pointer to the stepper motor controller instance
+ * @param steps_per_second Enter the speed in steps per second
+ * @return <0 Error during Enabling
+ *          0 Success
+ */
+__syscall int32_t stepper_set_velocity(const struct device *dev, uint32_t steps_per_second);
+
+static inline int32_t z_impl_stepper_set_velocity(const struct device *dev,
+						  uint32_t steps_per_second)
+{
+	const struct stepper_api *api = (const struct stepper_api *)dev->api;
+
+	return api->stepper_set_velocity(dev, steps_per_second);
+}
+
+/**
+ * @brief Set the microstep resolution in stepper motor controller
+ * @param dev pointer to the stepper motor controller instance
+ * @param resolution microstep resolution
+ * @return <0 Error during Enabling
+ *          0 Success
+ */
+__syscall int32_t stepper_set_micro_step_res(const struct device *dev,
+					     enum micro_step_resolution resolution);
+
+static inline int32_t z_impl_stepper_set_micro_step_res(const struct device *dev,
+							enum micro_step_resolution resolution)
+{
+	const struct stepper_api *api = (const struct stepper_api *)dev->api;
+
+	return api->stepper_set_micro_step_res(dev, resolution);
```

Comment:
I'd imagine some controller to have fixed microstepping and not being able to implement this, maybe add a

```c
if (api->stepper_set_micro_step_res) {
        return -ENOSYS;
}
```